core/tracker/tracker.go

```diff
@@ -164,34 +164,88 @@ func failedDutyReporter(ctx context.Context, duty core.Duty, failed bool, compon
 }
 
 // analyseParticipation returns a set of share indexes of participated peers.
-func analyseParticipation(events []event) (map[int]bool, error) {
+func analyseParticipation(ctx context.Context, duty core.Duty, allEvents map[core.Duty][]event) (map[int]bool, map[int]bool, error) {
 	// Set of shareIdx of participated peers.
 	resp := make(map[int]bool)
+	unexpectedPeers := make(map[int]bool)
 
-	for _, e := range events {
+	for _, e := range allEvents[duty] {
 		// If we get a parSigDBInternal event, then the current node participated successfully.
 		// If we get a parSigEx event, then the corresponding peer with e.shareIdx participated successfully.
-		if e.component == parSigEx || e.component == parSigDBInternal {
+		if e.component == parSigEx {
+			if err := validateParticipation(duty, e.pubkey, allEvents); err != nil {
+				log.Warn(ctx, "Unexpected event found", err, z.Int("ShareIdx", e.shareIdx))
+				unexpectedPeers[e.shareIdx] = true
+
+				continue
+			}
+
+			if e.shareIdx == 0 {
+				return nil, nil, errors.New("shareIdx empty", z.Any("component", e.component))
+			}
+			resp[e.shareIdx] = true
+		} else if e.component == parSigDBInternal {
 			if e.shareIdx == 0 {
-				return nil, errors.New("shareIdx empty", z.Any("component", e.component))
+				return nil, nil, errors.New("shareIdx empty", z.Any("component", e.component))
 			}
 			resp[e.shareIdx] = true
 		}
 	}
 
-	return resp, nil
+	return resp, unexpectedPeers, nil
+}
+
+// validateParticipation validates events from peers for a given duty.
+func validateParticipation(duty core.Duty, pubkey core.PubKey, allEvents map[core.Duty][]event) error {
```

Comment:
nit: `isParSigEventExpected return true if partially signed data events is expected for the given duty and pubkey. Partially signed data events are generated by parsigex and parsigdb.`

Reply:
done